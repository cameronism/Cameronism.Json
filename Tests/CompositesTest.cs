using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using Xunit;

namespace Tests
{
	public class CompositesTest
	{
		static Type GetType<T>(T sample)
		{
			return typeof(T);
		}

		[Fact]
		public void Reflect()
		{
			var sb = new StringBuilder();

			foreach (var t in new[] { 
				typeof(System.Int32),
				typeof(System.Boolean),
				typeof(System.Byte),
				typeof(System.Char),
				typeof(System.DateTime),
				typeof(System.Decimal),
				typeof(System.Double),
				typeof(System.Int16),
				typeof(System.Int64),
				typeof(System.SByte),
				typeof(System.Single),
				typeof(System.String),
				typeof(System.UInt16),
				typeof(System.UInt32),
				typeof(System.UInt64),

				typeof(System.Int32?),
				typeof(System.Boolean?),
				typeof(System.Byte?),
				typeof(System.Char?),
				typeof(System.DateTime?),
				typeof(System.Decimal?),
				typeof(System.Double?),
				typeof(System.Int16?),
				typeof(System.Int64?),
				typeof(System.SByte?),
				typeof(System.Single?),
				typeof(System.UInt16?),
				typeof(System.UInt32?),
				typeof(System.UInt64?),

				typeof(System.Guid),
				typeof(System.Guid?),

				//typeof(IEnumerable<int>),
				//typeof(string[]),
				//typeof(IReadOnlyList<Guid>),

				//GetType(new { i = 1 }),
				//GetType(new[] { new { s = "" } }.ToList()),

				//typeof(Dictionary<string, string>),
				//typeof(IReadOnlyDictionary<Guid, System.DayOfWeek>),

				//typeof(KeyValuePair<int, int>),
				//typeof(KeyValuePair<int, int>?),

				//typeof(DateTimeOffset),
				//typeof(DateTimeOffset?),

				//typeof(System.Net.IPAddress),

				//typeof(ExplicitDataMemberOrder),
			})
			{
				var schema = UnsafeJson.Schema.Reflect(t);
				DescribeMethod.MakeGenericMethod(t).Invoke(null, new object[] { sb });
			}

			ApprovalTests.Approvals.Verify(sb.ToString());
		}

		static readonly MethodInfo DescribeMethod = typeof(CompositesTest).GetMethod("Describe", BindingFlags.Static | BindingFlags.NonPublic);

		static void Describe<T>(StringBuilder sb)
		{
			sb.AppendLine();
			sb.AppendLine();
			sb.AppendLine("# " + SchemaTest.HumanName(typeof(T)));

			var emit = UnsafeJson.Composites.Create<T>(UnsafeJson.Schema.Reflect(typeof(T)));

			string instructions;
			var writer = emit.CreateDelegate(out instructions);
			sb.AppendLine(instructions);
		}
	}
}
