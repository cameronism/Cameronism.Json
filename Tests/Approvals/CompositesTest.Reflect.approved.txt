

# System.Int32
ldarg.0
ldind.i4
ldarg.1
ldarg.2
call Int32 WriteInt32(Int32, Byte*, Int32)
ret



# System.Boolean
ldarg.0
ldind.i1
ldarg.1
ldarg.2
call Int32 WriteBoolean(Boolean, Byte*, Int32)
ret



# System.Byte
ldarg.0
ldind.u1
conv.u4
ldarg.1
ldarg.2
call Int32 WriteUInt32(UInt32, Byte*, Int32)
ret



# System.Char
ldarg.0
ldind.u2
ldarg.1
ldarg.2
call Int32 WriteCharUtf8(Char, Byte*, Int32)
ret



# System.DateTime
ldarg.0
ldobj System.DateTime
ldarg.1
ldarg.2
call Int32 WriteDateTime8601(System.DateTime, Byte*, Int32)
ret



# System.Decimal
ldarg.0
ldobj System.Decimal
ldarg.1
ldarg.2
call Int32 WriteDecimal(System.Decimal, Byte*, Int32)
ret



# System.Double
ldarg.0
ldind.r8
ldarg.1
ldarg.2
call Int32 WriteDouble(Double, Byte*, Int32)
ret



# System.Int16
ldarg.0
ldind.i2
conv.i4
ldarg.1
ldarg.2
call Int32 WriteInt32(Int32, Byte*, Int32)
ret



# System.Int64
ldarg.0
ldind.i8
ldarg.1
ldarg.2
call Int32 WriteInt64(Int64, Byte*, Int32)
ret



# System.SByte
ldarg.0
ldind.i1
conv.i4
ldarg.1
ldarg.2
call Int32 WriteInt32(Int32, Byte*, Int32)
ret



# System.Single
ldarg.0
ldind.r4
ldarg.1
ldarg.2
call Int32 WriteSingle(Single, Byte*, Int32)
ret



# System.String
ldarg.0
ldind.ref
ldarg.1
ldarg.2
call Int32 WriteStringUtf8(System.String, Byte*, Int32)
ret



# System.UInt16
ldarg.0
ldind.u2
conv.u4
ldarg.1
ldarg.2
call Int32 WriteUInt32(UInt32, Byte*, Int32)
ret



# System.UInt32
ldarg.0
ldind.u4
ldarg.1
ldarg.2
call Int32 WriteUInt32(UInt32, Byte*, Int32)
ret



# System.UInt64
ldarg.0
ldind.i8
ldarg.1
ldarg.2
call Int32 WriteUInt64(UInt64, Byte*, Int32)
ret



# System.Nullable<System.Int32>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call Int32 GetValueOrDefault()
ldarg.1
ldarg.2
call Int32 WriteInt32(Int32, Byte*, Int32)
ret



# System.Nullable<System.Boolean>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call Boolean GetValueOrDefault()
ldarg.1
ldarg.2
call Int32 WriteBoolean(Boolean, Byte*, Int32)
ret



# System.Nullable<System.Byte>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call Byte GetValueOrDefault()
conv.u4
ldarg.1
ldarg.2
call Int32 WriteUInt32(UInt32, Byte*, Int32)
ret



# System.Nullable<System.Char>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call Char GetValueOrDefault()
ldarg.1
ldarg.2
call Int32 WriteCharUtf8(Char, Byte*, Int32)
ret



# System.Nullable<System.DateTime>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call System.DateTime GetValueOrDefault()
ldarg.1
ldarg.2
call Int32 WriteDateTime8601(System.DateTime, Byte*, Int32)
ret



# System.Nullable<System.Decimal>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call System.Decimal GetValueOrDefault()
ldarg.1
ldarg.2
call Int32 WriteDecimal(System.Decimal, Byte*, Int32)
ret



# System.Nullable<System.Double>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call Double GetValueOrDefault()
ldarg.1
ldarg.2
call Int32 WriteDouble(Double, Byte*, Int32)
ret



# System.Nullable<System.Int16>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call Int16 GetValueOrDefault()
conv.i4
ldarg.1
ldarg.2
call Int32 WriteInt32(Int32, Byte*, Int32)
ret



# System.Nullable<System.Int64>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call Int64 GetValueOrDefault()
ldarg.1
ldarg.2
call Int32 WriteInt64(Int64, Byte*, Int32)
ret



# System.Nullable<System.SByte>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call SByte GetValueOrDefault()
conv.i4
ldarg.1
ldarg.2
call Int32 WriteInt32(Int32, Byte*, Int32)
ret



# System.Nullable<System.Single>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call Single GetValueOrDefault()
ldarg.1
ldarg.2
call Int32 WriteSingle(Single, Byte*, Int32)
ret



# System.Nullable<System.UInt16>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call UInt16 GetValueOrDefault()
conv.u4
ldarg.1
ldarg.2
call Int32 WriteUInt32(UInt32, Byte*, Int32)
ret



# System.Nullable<System.UInt32>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call UInt32 GetValueOrDefault()
ldarg.1
ldarg.2
call Int32 WriteUInt32(UInt32, Byte*, Int32)
ret



# System.Nullable<System.UInt64>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call UInt64 GetValueOrDefault()
ldarg.1
ldarg.2
call Int32 WriteUInt64(UInt64, Byte*, Int32)
ret



# System.Guid
ldarg.0
ldobj System.Guid
ldarg.1
ldarg.2
call Int32 WriteGuidFormatD(System.Guid, Byte*, Int32)
ret



# System.Nullable<System.Guid>
ldarg.0
dup
call Boolean get_HasValue()
brtrue.s HasValue_true
pop
ldarg.1
ldarg.2
call Int32 WriteNull(Byte*, Int32)
ret

HasValue_true:
call System.Guid GetValueOrDefault()
ldarg.1
ldarg.2
call Int32 WriteGuidFormatD(System.Guid, Byte*, Int32)
ret

